reactor-core/src/test/java/reactor/core/publisher/MonoCollectTest.java

```diff
@@ -140,4 +143,59 @@ public void scanSubscriber() {
 		assertThat(test.scan(Scannable.Attr.CANCELLED)).isTrue();
 	}
 
+	@Test
+	public void discardOnError() {
+		List<Integer> list = new ArrayList<>();
+		list.add(1);
+		list.add(2);
+		list.add(3);
+		list.add(4);
+
+		AtomicBoolean res = new AtomicBoolean();
+
+		Mono<List<Integer>> test = Flux.range(1, 10)
+		                               .hide()
+		                               .map(i -> {
+			                               if (i == 5) {
+				                               throw new IllegalStateException("boom");
+			                               }
+			                               return i;
+		                               }).<List<Integer>>collect(ArrayList::new,
```

Comment:
nit: I would put `.collect` and `.doOnDiscard` to the separate lines, for readability (also in `discardOnCancel`)